### PR TITLE
feat(cms): showing online users

### PR DESCRIPTION
**Pull request by nickhsine**

### CODE REVIEW

**Comment by duidae** [comment]
@nickhsine cant assign reviewer to you 😆, please have a look, thanks

get current user info from `authenticatedItem`

Screen Shot 2023-11-27 at 1 52 28 PM

File: packages/cms/lists/views/online-users.tsx [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            isCurrentUserExisting
+              ? [...users]
+              : [
+                  ...users,
+                  {
+                    email: currentUserEmail,
+                    name: authenticatedItem.name,
+                    id: authenticatedItem.id,
+                  },
+                ]
```
Comment (nickhsine) [inline packages/cms/lists/views/online-users.tsx:193]: Question:
如果 user 的 role 沒有 update 的權限，是否就不需要顯示該使用者？

建議修改：
```javascript
const updateResults = await axios.post('/api/graphql', {
  query: upateUserGql,
  ...
})

// error to update `onlineUsers`, maybe the authenticated user does not hav e update permission
if (updateResults?.data?.errors) {
  // do nothing
} else if (!isCurrentUserExisting) {
  setUsers([...users, { email: ... }])
}
```

**Comment by nickhsine** [comment]
LGTM 👍 👍 
Only some suggestions remains.

File: packages/cms/lists/views/online-users.tsx [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+`
+
+type User = {
+  id: string
+  name: string
+  email: string
+}
+
+export const Field = ({ value }: FieldProps<typeof controller>) => {
+  let currentUserEmail = ''
```
Comment (nickhsine) [inline packages/cms/lists/views/online-users.tsx]: Suggestion:
Maybe `useRef` to replace js closure.

File: packages/cms/lists/views/online-users.tsx [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  // Polling & clear polling/current user field when leaving
+  useEffect(() => {
+    const polling = setInterval(async () => {
+      const users = await handleQueryUsers()
+      setUsers(users ?? [])
+    }, 5000)
+
+    return () => {
+      clearInterval(polling)
+      handleRemoveUser()
```
Comment (nickhsine) [inline packages/cms/lists/views/online-users.tsx:214]: Suggestion:
如果使用者沒有權限 update `onlineUsers` 欄位的話，是不是可以省略 `handleRemoveUser()`？

File: packages/cms/lists/views/online-users.tsx (hunk 2 above)
Comment (duidae) [reply packages/cms/lists/views/online-users.tsx]: fixed.

File: packages/cms/lists/views/online-users.tsx (hunk 3 above)
Comment (duidae) [reply packages/cms/lists/views/online-users.tsx:214]: fixed.

File: packages/cms/lists/views/online-users.tsx (hunk 1 above)
Comment (duidae) [reply packages/cms/lists/views/online-users.tsx:193]: fixed.

**Comment by duidae** [comment]
> LGTM 👍 👍 Only some suggestions remains.

@nickhsine comments are fixed as above, please have a look, thanks!

File: packages/cms/lists/views/online-users.tsx [hunk 4]
```diff
@@ -175,18 +176,22 @@ export const Field = ({ value }: FieldProps<typeof controller>) => {
...
+            isCurrentUserWritable.current && !isCurrentUserExisting
+              ? [
                   ...users,
                   {
                     email: currentUserEmail.current,
                     name: authenticatedItem.name,
                     id: authenticatedItem.id,
                   },
                 ]
+              : [...users]
```
Comment (nickhsine) [inline packages/cms/lists/views/online-users.tsx:194]: Question:
if `users` is not changed, do we need to `setUsers([...users])` again to trigger re-render?

**Review by nickhsine — COMMENTED** [review]
LGTM 👍 👍 
Just one question left.

File: packages/cms/lists/views/online-users.tsx (hunk 4 above)
Comment (duidae) [reply packages/cms/lists/views/online-users.tsx:194]: this useeffect is a constructor, will not be triggered after didmount, and `users` state is empty in the beginning

Comment (nickhsine) [reply packages/cms/lists/views/online-users.tsx:194]: Got it!
謝謝說明，我沒有注意到這個 useEffect 只會跑 didMount 那一次。

**Pull request merged**